composite/token-allocation/src/adapter.ts

```diff
@@ -35,19 +36,34 @@ export const marketCapTotalValue = (
     .toNumber()
 }
 
-const inputParams = {
-  allocations: true,
-  quote: false,
-}
-
 const toValidAllocations = (allocations: TokenAllocations): TokenAllocations => {
   if (!allocations.every((t) => !!t.symbol))
     throw new AdapterError({ message: `Symbol not available for all tokens.`, statusCode: 400 })
-  return allocations.map((t) => ({
-    symbol: t.symbol.toUpperCase(),
-    decimals: t.decimals || DEFAULT_TOKEN_DECIMALS,
-    balance: t.balance || DEFAULT_TOKEN_BALANCE * 10 ** (t.decimals || DEFAULT_TOKEN_DECIMALS),
-  }))
+
+  const _toValidDecimals = (decimals: number) =>
+    Number.isInteger(decimals) && decimals >= 0 ? decimals : DEFAULT_TOKEN_DECIMALS
+  const _toValidBalance = (balance: BigNumberish, decimals: number) => {
+    if (!balance) {
+      return DEFAULT_TOKEN_BALANCE * 10 ** decimals
+    }
+    if (!Number(balance)) {
+      throw new AdapterError({
+        message: `Invalid balance`,
+        statusCode: 400,
+      })
+    }
+    if (Number(balance) < 0)
```

Comment:
I don't think this is going to work with balance being `BigNumber`?

You'll need to convert from `BigNumberish` to `number`:

```suggestion
    if (BigNumber.from(balance).toNumber() < 0)
```

Reply:
The input types are wrong, should be:

```javascript
const _toValidBalance = (balance: number | string | undefined, decimals: number) => {
```

Reply:
Accepting only `number | string | undefined` will have some conflicts, as `TokenAllocations` (type) accepts balance as `BigNumberish`.

Reply:
```
type BigNumberish = BigNumber | Bytes | string | number;
```

`_toValidBalance` is happening before we have `TokenAllocations`. We are trying to map and validate the `number | string | undefined` input as `BigNumber`.

Reply:
> Also, what are you exactly checking here? When do you expect to throw "Invalid balance"?

Anything that is not `BigNumberish`. This could be thrown if the `TokenAllocations` adapter is used directly, and the client gives some invalid balance.

Reply:
The easiest way is to check for `NaN` like so: `if (isNaN(Number(balance)) || Number(balance) < 0) throw`